src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -1157,11 +1157,13 @@ private bool IsValidDataExtensionPropertyType(ITypeSymbol type)
                 }
 
                 if ((!canUseGetter && !canUseSetter && !hasJsonIncludeButIsInaccessible) ||
-                    !IsSymbolAccessibleWithin(memberType, within: contextType))
+                    !IsSymbolAccessibleWithin(memberType, within: contextType) ||
+                    (memberType.IsRefLikeType && ignoreCondition is JsonIgnoreCondition.Always))
                 {
                     // Skip the member if either of the two conditions hold
                     // 1. Member has no accessible getters or setters (but is not marked with JsonIncludeAttribute since we need to throw a runtime exception) OR
-                    // 2. The member type is not accessible within the generated context.
+                    // 2. The member type is not accessible within the generated context OR
+                    // 3. The member is of a ref-like type that is also ignored (we keep all other ignored members for backward compat)
```

Comment:
Why don't you ignore all ref-like properties regardless of [JsonIgnore]? Those would still cause errors. Plus, this won't work for types not owned by the user, as is the case of https://github.com/dotnet/runtime/issues/98590#issuecomment-1950299008.

Reply:
Good question, the reflection-based serializer will throw an exception for types that define un-ignored ref-like members, so for the sake of consistency we don't want to silently skip those even if that might result in a compile-time error. What we might be able to do though is emit a diagnostic (warning or error) instead of generating uncompilable code.

Reply:
I've updated the changes to always skip ref like properties and constructors, but to also throw a warning diagnostic if they aren't being ignored. I've also updated both the source generator and reflection serializer to include handling for collection types whose elements are ref-like elements, something that is possible starting with C# 13.